### ISSUE TITLE
Correctly tracking inf and nan with a double type

### ISSUE THREAD

**Issue by josyoun** [dhall-lang/dhall-lang#537]
Is there a chance that we will eventually correctly and explicitely track inf and nan with the double type?  This is a follow-up to a previously filed issue #252.  Within that issue, a decision was made to represent inf the same as jq.  While this is pragmatic, I contend this is the wrong representation given that there are other besides json.

### Why this is important

- Configuration files for mathematical programs require parameters in IEEE Standard 754 format

- inf is useful for representing parameters without bound

- nan is useful for representing invalid, uninitialized values

### Problems with current implementation

- jq represents inf as realmax, but these are not the same and this will cause errors.  A common test for inf is whether or not the value is larger than real max.  For example, the following works in Octave, but a similar example works in C
```
octave:2> inf > realmax
ans = 1
octave:3> realmax > inf
ans = 0
```
This means that the current output from dhall when read into another program will give a fundamentally incorrect answer to these tests.

- Unlike JSON, YAML correctly allows a [representation](https://yaml.org/type/float.html) of both nan and inf.  However, the output from dhall does not match these specification.  For example, the dhall program
```
[NaN,Infinity,-Infinity]
```
Produces
```
$ cat test02.dhall | dhall-to-yaml 
- null
- 179769313486231570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
- -179769313486231570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```
This in incorrect, I contend it should be
```
-  NaN
- Inf
- -Inf
```

### Problems with changing

- Candidly, this is a pain because, fundamentally, JSON does not allow inf or nan to be represented directly.  This is a problem with JSON and an ongoing source of aggravation for people like me who need to directly encode nan and inf into configuration files.

- Certainly, it's possible to have the YAML output differ from JSON, but then they will be different, which makes consistency more difficult to maintain

### Proposed changes

- Although jq returns realmax for inf, I contend this is an incorrect and potentially code breaking decision that dhall should not mirror.  It's deceptive and not the same.  Generally, when I have to encode inf and nan in JSON, the safe way to handle this is to encode them as the strings "inf" and "nan".  This potentially makes the client code using the JSON file work harder because it needs to check the type before extracting the value, but at least this is safe.  Returning realmax is not safe.  It's kind of like returning a 2 instead of 1.

- Have the YAML output return NaN, Inf, and -Inf as the specification allows

### Future benefits

- If dhall decides to expand the floating point arithmatic operators, treating nan and inf correctly will already be complete.  Even if further arithmetic operations are not supported, representing these values as literals is still very important for some applications.

- If dhall becomes more of a direct replacement for JSON and YAML, a missing feature of JSON will be corrected and a nice feature of YAML will be preserved

**Comment by singpolyma** [comment]
This seems dhall-haskell (and possible dhall-json) specific.

**Comment by Nadrieril** [comment]
I believe this is a problem with the Haskell implementation. The standard doesn't specify how Doubles should be represented internally. In fact I would say that the name `Infinity` strongly hints at representing `Infinity` by the float infinity value rather than anything else.

**Comment by Gabriel439** [comment]
Yeah, I will migrate this issue to the `dhall-haskell` repository since it seems specific to the Haskell implementation
